fish-rust/src/wutil/mod.rs

```diff
@@ -32,6 +82,402 @@ pub fn perror(s: &str) {
+    assert_eq!(norm_path(L!("///")), "/");
+    assert_eq!(norm_path(L!("////")), "/");
+    assert_eq!(norm_path(L!("/.///")), "/");
+    assert_eq!(norm_path(L!(".//")), ".");
+    assert_eq!(norm_path(L!("/.//../")), "/");
+    assert_eq!(norm_path(L!("////abc")), "/abc");
+    assert_eq!(norm_path(L!("/abc")), "/abc");
+    assert_eq!(norm_path(L!("/abc/")), "/abc");
+    assert_eq!(norm_path(L!("/abc/..def/")), "/abc/..def");
+    assert_eq!(norm_path(L!("//abc/../def/")), "//def");
+    assert_eq!(norm_path(L!("abc/../abc/../abc/../abc")), "abc");
+    assert_eq!(norm_path(L!("../../")), "../..");
+    assert_eq!(norm_path(L!("foo/./bar")), "foo/bar");
+    assert_eq!(norm_path(L!("foo/../")), ".");
+    assert_eq!(norm_path(L!("foo/../foo")), "foo");
+    assert_eq!(norm_path(L!("foo/../foo/")), "foo");
+    assert_eq!(norm_path(L!("foo/././bar/.././baz")), "foo/baz");
+}
+
+/// Given an input path \p path and a working directory \p wd, do a "normalizing join" in a way
+/// appropriate for cd. That is, return effectively wd + path while resolving leading ../s from
+/// path. The intent here is to allow 'cd' out of a directory which may no longer exist, without
+/// allowing 'cd' into a directory that may not exist; see #5341.
+#[widestrs]
+pub fn path_normalize_for_cd(wd: &wstr, path: &wstr) -> WString {
+    // Fast paths.
+    const sep: char = '/';
+    assert!(
+        wd.as_char_slice().first() == Some(&'/') && wd.as_char_slice().last() == Some(&'/'),
+        "Invalid working directory, it must start and end with /"
+    );
+    if path.is_empty() {
+        return wd.to_owned();
+    } else if path.as_char_slice().first() == Some(&sep) {
+        return path.to_owned();
+    } else if path.as_char_slice().first() != Some(&'.') {
+        return wd.to_owned() + path;
+    }
+
+    // Split our strings by the sep.
+    let mut wd_comps = split_string(wd, sep);
+    let path_comps = split_string(path, sep);
+
+    // Remove empty segments from wd_comps.
+    // In particular this removes the leading and trailing empties.
+    wd_comps.retain(|comp| !comp.is_empty());
+
+    // Erase leading . and .. components from path_comps, popping from wd_comps as we go.
+    let mut erase_count = 0;
+    for comp in &path_comps {
+        let mut erase_it = false;
+        if comp.is_empty() || comp == "."L {
+            erase_it = true;
+        } else if comp == ".."L && !wd_comps.is_empty() {
+            erase_it = true;
+            wd_comps.pop();
+        }
+        if erase_it {
+            erase_count += 1;
+        } else {
+            break;
+        }
+    }
+    // Append un-erased elements to wd_comps and join them, then prepend the leading /.
+    wd_comps.extend(path_comps.into_iter().skip(erase_count));
+
+    let mut result = join_strings(&wd_comps, sep);
+    result.insert(0, '/');
+    result
+}
+
+/// Wide character version of dirname().
+#[widestrs]
+pub fn wdirname(path: WString) -> WString {
```

Comment:
In general, when we're dealing with paths and wstrings, I have a question:

How do these work with NULs? Because UNIX paths can't have NULs, so this should obey c-string semantics and only operate until the first NUL.

Reply:
yeah that's an interesting case. While `wdirname()` accepts interior NULs, an eventual system call will not.
Rust makes things better because CString will report interior NULs as error instead of std::string's silent truncation.

For now I maintained the existing behavior in the place where most CStrings are created:

```rust
pub fn wcs2zstring(input: &wstr) -> CString {
    if input.is_empty() {
        return CString::default();
    }

    let mut result = vec![];
    wcs2string_callback(input, |buff| {
        result.extend_from_slice(buff);
        true
    });
    let until_nul = match result.iter().position(|c| *c == b'\0') {
        Some(pos) => &result[..pos],
        None => &result[..],
    };
    CString::new(until_nul).unwrap()
}
```

which is of course super surprising (in combination with nul-agnostic `wdirname` and such) but it has always been that way and not caused problems